This folder contains a command-line Pythonic program to filter a VCF file, using cyvcf2, while parallelizing filtration across each chromosome to dramatically increase the speed of filtration.

The program is contained in filter_vcf.py and the run_filter_vcf.sh file is an example (and editable) Bash script that defines all its input parameters and runs it. In this case, the entire program can be run by just executing run_filter_vcf.sh in the command-line.

There are two additional files that may go alongside this program (other than the input VCF, of course). One is a file that specifies the regions you want to parallelize filtration across. In this directory, it is called chromosomes.txt. It is newline-separated, specifying each region that filtration must be done on per line. This file is required. The second file is optional but allows for filtration to be done on in relation to the average sequencing depth of each individual in the VCF. For example, it may be desired to filter sites where many individuals have a sequencing depth above twice the average to exclude regions with copy number variation.
